polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -498,27 +466,57 @@ where
 		rotation_info.core_for_group(group, cores.len())
 	} else {
 		gum::trace!(target: LOG_TARGET, ?relay_parent, "Not a validator");
-		return Ok(())
+		return Ok(GroupAssignments { current: Vec::new() })
 	};
 
-	let paras_now = match fetch_claim_queue(sender, relay_parent).await.map_err(Error::Runtime)? {
-		// Runtime supports claim queue - use it
-		//
-		// `relay_parent_mode` is not examined here because if the runtime supports claim queue
-		// then it supports async backing params too (`ASYNC_BACKING_STATE_RUNTIME_REQUIREMENT`
-		// < `CLAIM_QUEUE_RUNTIME_REQUIREMENT`).
-		Some(mut claim_queue) => claim_queue.0.remove(&core_now),
-		// Claim queue is not supported by the runtime - use availability cores instead.
-		None => cores.get(core_now.0 as usize).and_then(|c| match c {
-			CoreState::Occupied(core) if relay_parent_mode.is_enabled() =>
-				core.next_up_on_available.as_ref().map(|c| [c.para_id].into_iter().collect()),
-			CoreState::Scheduled(core) => Some([core.para_id].into_iter().collect()),
-			CoreState::Occupied(_) | CoreState::Free => None,
-		}),
-	}
-	.unwrap_or_else(|| VecDeque::new());
+	// Assigned parachains are determined by:
+	// 1. If there is something scheduled on the core - this is the first assignment.
+	// 2. Next assigments are whatever is in the claim queue for the core. 2.1 If the claim queue is
+	//    not available and the core is occupied the next assignemnt is taken from
+	//    `next_up_on_available`
+	//
+	// Step 1 is needed to unify the assignments with the way backing and prospective parachains
+	// subsystems handle async backing parameters. Until
+	// https://github.com/paritytech/polkadot-sdk/issues/5079 is completed the current view consists
+	// of the latest known relay parent plus its `lookahead` number of ancestors. When #5079 is
+	// merged everything will be deduced from the claim queue and this logic will be simplified.
+	let assigned_paras = {
+		let mut from_core: VecDeque<_> =
+			if let Some(CoreState::Scheduled(core_state)) = cores.get(core_now.0 as usize) {
+				vec![core_state.para_id].into_iter().collect()
+			} else {
+				vec![].into_iter().collect()
+			};
+
+		let from_claim_queue =
+			match fetch_claim_queue(sender, relay_parent).await.map_err(Error::Runtime)? {
+				// Runtime supports claim queue - use it
+				Some(mut claim_queue) => claim_queue.0.remove(&core_now),
+				// Should never happen since claim queue is released everywhere.
+				None => {
+					gum::warn!(
+						target: LOG_TARGET,
+						?relay_parent,
+						"Claim queue is not available, assigning from availability cores only",
+					);
+
+					if let Some(CoreState::Occupied(core_state)) = cores.get(core_now.0 as usize) {
+						core_state.next_up_on_available.as_ref().map(|scheduled_core| {
+							vec![scheduled_core.para_id].into_iter().collect()
+						})
+					} else {
+						None
+					}
```

Comment:
I'm not sure if we still need to handle missing claim queue or handle it as an error (the runtime api is released everywhere). Since it's a small addition I decided to keep it here.

Reply:
To simplify the code there will be no fallback for missing claim queue runtime api.